tfhe/src/strings/test_functions/test_concat.rs

```diff
@@ -1,157 +1,146 @@
+use crate::integer::keycache::KEY_CACHE;
+use crate::integer::server_key::radix_parallel::tests_cases_unsigned::FunctionExecutor;
+use crate::integer::server_key::radix_parallel::tests_unsigned::CpuFunctionExecutor;
+use crate::integer::{IntegerKeyKind, RadixClientKey, ServerKey};
 use crate::shortint::parameters::PARAM_MESSAGE_2_CARRY_2_KS_PBS_TUNIFORM_2M64;
-use crate::strings::ciphertext::UIntArg;
-use crate::strings::test::TestKind;
-use crate::strings::test_functions::result_message_rhs;
-use crate::strings::TestKeys;
-use std::time::Instant;
+use crate::shortint::PBSParameters;
+use crate::strings::ciphertext::{FheString, UIntArg};
+use std::sync::Arc;
 
 const TEST_CASES_CONCAT: [&str; 5] = ["", "a", "ab", "abc", "abcd"];
 
 #[test]
-fn test_concat_trivial() {
-    let keys = TestKeys::new(
-        PARAM_MESSAGE_2_CARRY_2_KS_PBS_TUNIFORM_2M64,
-        TestKind::Trivial,
-    );
+fn string_concat_test_parameterized() {
+    string_concat_test(PARAM_MESSAGE_2_CARRY_2_KS_PBS_TUNIFORM_2M64);
+}
+
+#[allow(clippy::needless_pass_by_value)]
+fn string_concat_test<P>(param: P)
+where
+    P: Into<PBSParameters>,
+{
+    let executor = CpuFunctionExecutor::new(&ServerKey::concat);
+    string_concat_test_impl(param, executor, |lhs, rhs| lhs.to_owned() + rhs);
+}
 
+pub(crate) fn string_concat_test_impl<P, T>(
+    param: P,
+    mut concat_executor: T,
+    clear_function: for<'a> fn(&'a str, &'a str) -> String,
```

Comment:
Actually, there is no need for the clear_function in any of the tests

The test_impl function is what does the test, and so it knows what the output is expected to be

For example,  If the caller of the test could give concat_executor and a clear_function that do the same things but are not concatenating

Reply:
Many tests test more than one function

On the tests that only test one function, I could remove it.
But not sure it's better as there will be 2 kinds of test structures instead of one.
What do you think?

Reply:
Maybe you could do it like its done for some of the tests: the test functions receive multiple executor:

```rust
pub(crate) fn string_contains_test_impl<P, E1, E2, E3>(
    param: P,
    mut contains_executor: E1,
    mut starts_with_executor: E2,
    mut end_with_executor: E3,
) {
 //..
}
```
what do you think ?